server/backup_binlog.go

```diff
@@ -0,0 +1,346 @@
+	"github.com/cybozu-go/moco"
+	mocoagent "github.com/cybozu-go/moco-agent"
+	"github.com/cybozu-go/well"
+	"github.com/jmoiron/sqlx"
+)
+
+// BackupBinaryLogsParams is the paramters for backup binary logs
+type BackupBinaryLogsParams struct {
+	FilePrefix   string
+	BucketHost   string
+	BucketPort   int
+	BucketName   string
+	BucketRegion string
+
+	AccessKeyID     string
+	SecretAccessKey string
+}
+
+// FlushAndBackupBinaryLogs executes "FLUSH BINARY LOGS;"
+// and upload it to the object storage, then delete it
+func (a *Agent) FlushAndBackupBinaryLogs(w http.ResponseWriter, r *http.Request) {
+	var err error
+	if r.Method != http.MethodGet {
+		w.WriteHeader(http.StatusMethodNotAllowed)
+		return
+	}
+
+	token := r.URL.Query().Get(moco.AgentTokenParam)
+	if token != a.token {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	params, err := parseBackupBinLogParams(r.URL.Query())
+	if err != nil {
+		w.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	if !a.sem.TryAcquire(1) {
+		w.WriteHeader(http.StatusTooManyRequests)
+		return
+	}
+
+	sess := session.Must(session.NewSession(&aws.Config{
+		Region:           aws.String(params.BucketRegion),
+		Endpoint:         aws.String(fmt.Sprintf("%s:%d", params.BucketHost, params.BucketPort)),
+		DisableSSL:       aws.Bool(true),
+		S3ForcePathStyle: aws.Bool(true),
+		Credentials:      credentials.NewStaticCredentials(params.AccessKeyID, params.SecretAccessKey, ""),
+	}))
+
+	// prevent re-execution of the same request
+	_, err = s3.New(sess).HeadObject(&s3.HeadObjectInput{
+		Bucket: aws.String(params.BucketName),
+		Key:    aws.String(params.FilePrefix),
+	})
+	if err == nil {
+		a.sem.Release(1)
+		w.WriteHeader(http.StatusConflict)
+		return
+	}
+	if strings.HasPrefix(err.Error(), s3.ErrCodeNoSuchKey) {
+		a.sem.Release(1)
+		internalServerError(w, fmt.Errorf("failed to get objects: %w", err))
+		log.Error("failed to get objects", map[string]interface{}{
+			log.FnError: err,
+		})
+		return
+	}
+
+	// TODO: change user
+	rootPassword := os.Getenv(moco.RootPasswordEnvName)
+	db, err := a.acc.Get(fmt.Sprintf("%s:%d", a.mysqlAdminHostname, a.mysqlAdminPort), moco.RootUser, rootPassword)
+	if err != nil {
+		a.sem.Release(1)
+		internalServerError(w, fmt.Errorf("failed to connect to database before flush binary logs: %w", err))
+		log.Error("failed to connect to database before flush binary logs", map[string]interface{}{
+			"hostname":  a.mysqlAdminHostname,
+			"port":      a.mysqlAdminPort,
+			log.FnError: err,
+		})
+		return
+	}
+
+	err = flushBinaryLogs(r.Context(), db)
+	if err != nil {
+		a.sem.Release(1)
+		internalServerError(w, fmt.Errorf("failed to flush binary logs: %w", err))
+		log.Error("failed to flush binary logs", map[string]interface{}{
+			log.FnError: err,
+		})
+		return
+	}
+
+	env := well.NewEnvironment(context.Background())
+	env.Go(func(ctx context.Context) error {
+		defer a.sem.Release(1)
+
+		err = uploadBinaryLog(r.Context(), sess, db, params)
```

Comment:
I think you should use `ctx` instead of `r.Context()`.
`r.Context()` will not be canceled at the end of the request?